rbac_api.go

```diff
@@ -51,36 +50,76 @@ func (e *Enforcer) HasRoleForUser(name string, role string) (bool, error) {
 
 // AddRoleForUser adds a role for a user.
 // Returns false if the user already has the role (aka not affected).
-func (e *Enforcer) AddRoleForUser(user string, role string) (bool, error) {
-	return e.AddGroupingPolicy(user, role)
+func (e *Enforcer) AddRoleForUser(user string, role string, domain ...string) (bool, error) {
+	if len(domain) ==0 {
+		return e.AddGroupingPolicy(user, role)
+	}else if len(domain) > 1 {
+		return false, errors.ERR_DOMAIN_PARAMETER
+	}
+	return e.AddGroupingPolicy(user, role, domain[0])
 }
 
 // AddRolesForUser adds roles for a user.
 // Returns false if the user already has the roles (aka not affected).
-func (e *Enforcer) AddRolesForUser(user string, roles []string) (bool, error) {
+func (e *Enforcer) AddRolesForUser(user string, roles []string, domain ...string) (bool, error) {
 	f := false
-	for _, r := range roles {
-		b, err := e.AddGroupingPolicy(user, r)
-		if err != nil {
-			return false, err
+	if len(domain) == 0 {
+		for _, r := range roles {
+			b, err := e.AddGroupingPolicy(user, r)
+			if err != nil {
+				return false, err
+			}
+			if b {
+				f = true
+			}
 		}
-		if b {
-			f = true
+	}else if len(domain) > 1 {
+		return false,errors.ERR_DOMAIN_PARAMETER
+	}else {
+		for _, r := range roles {
+			b, err := e.AddGroupingPolicy(user, r, domain[0])
+			if err != nil {
+				return false, err
+			}
+			if b {
+				f = true
+			}
 		}
 	}
 	return f, nil
 }
 
 // DeleteRoleForUser deletes a role for a user.
 // Returns false if the user does not have the role (aka not affected).
-func (e *Enforcer) DeleteRoleForUser(user string, role string) (bool, error) {
-	return e.RemoveGroupingPolicy(user, role)
+func (e *Enforcer) DeleteRoleForUser(user string, role string, domain ...string) (bool, error) {
+	if len(domain) ==0 {
+		return e.RemoveGroupingPolicy(user, role)
+	}else if len(domain) > 1 {
+		return false, errors.ERR_DOMAIN_PARAMETER
+	}
+	return e.RemoveGroupingPolicy(user, role, domain[0])
 }
 
 // DeleteRolesForUser deletes all roles for a user.
 // Returns false if the user does not have any roles (aka not affected).
-func (e *Enforcer) DeleteRolesForUser(user string) (bool, error) {
-	return e.RemoveFilteredGroupingPolicy(0, user)
+func (e *Enforcer) DeleteRolesForUser(user string, domain ...string) (bool, error) {
+	if len(domain) == 0 {
+		return e.RemoveFilteredGroupingPolicy(0, user)
+	}else if len(domain) > 1 {
+		return false,errors.ERR_DOMAIN_PARAMETER
+	}
+
+	roles, err := e.model["g"]["g"].RM.GetRoles(user, domain[0])
+	if err != nil {
+		return false, err
+	}
+
+	var rules [][]string
+	for _, role := range roles {
+		rules = append(rules, []string{user, role, domain[0]})
+	}
+
+	return e.RemoveGroupingPolicies(rules)
```

Comment:
Why change the logic here?

Reply:
Beceuse if I use `e.RemoveFilteredGroupingPolicy(0, user, domain[0])`, there is a little problem.
***.csv
```
g, bob, admin, domain2
```
So the Policy is `bob, admin, domain2`. But the `fieldValues` is `bob, domain2`. It will not delete correctly.

Reply:
fieldValues supports to use empty string "" to skip some values in the middle.

Reply:
I will try

Reply:
updated